### DISCUSSION TITLE
Tips on Doctr Speed / Performance

### DISCUSSION THREAD

**Discussion by jhogendorn** [mindee/doctr#1094] [General]
I'm running the PDF/A notebook in a script pretty much copy and paste verbatim.

On an old mac laptop with an nvidia card but no proper cuda, i've got some benchmarks to OCR a 2 page document:

Library | Environment | CPU | GPU | Time
-------| -----|-----| -----| ------
Tensorflow | Python3.10 | X | | ~90s
Pytorch | Python3.10 | X | | ~80s 
Tensorflow | Anaconda | X | | ~30s
Pytorch | Anaconda | X | | ~25s
Tensorflow | Anaconda | | X | ~15s
Pytorch | Anaconda | | X | ~18s

Aside from a better machine/cuda support, what are the best optimisations we can make in the user land for speeding up inference with existing models?

Conda is a big speedup, I'm not 100% sure why. GPU usage is faster but less reliable, i run into issues where it has memory problems.

There are various articles on speedups for pytorch and tensorflow and I've tried some of them to not much success, as i don't deal with ML libs/processes normally.

Some options:

- Compiling pytorch on the machine?
- Configuration changes to reduce unnecessary work
- Installing supporting libraries to enable speedup features? (tensorflow-macos tensorflow-metal for example)
- Optimising the models for my machine somehow?
- Converting the models to tflite and trying to get it working with a Coral TPU or similar?
- My document source already deskews. I'm scanning at quite high quality, is there a data prep change that would help optimisation? ie would less dpi actually help rather than hinder.

Some guidance/discussion on whether these optimisations are worthwhile would be good also. IE, if i would be better served just switching to a cloud tpu or throwing a cheap 3 generation old nvidia card into a server here and passing it through to a docker container of my code, or something similar.

**Comment by felixdittrich92** [comment]
Hi @jhogendorn :wave:,

really interesting benchmark with conda and pip installations :+1: 
I am not a specialist in depth with conda and not 100% sure but maybe it has anything to do with the precompiled binaries which conda creates. (@frgfm Any idea how this can be possible ? :sweat_smile: )

In a first step i would suggest to try another model where the backbone is made to run on low resource systems like mobile phones or rasperry pi`s:

`model = ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch='crnn_mobilenet_v3_small', pretrained=True)  # if cuda available and you use PyTorch add .cuda()`

I took a while ago a short benchmark with this combination on an RTX3080 (mobile) comparable with the RTX2080 (desktop) ref. #876 (answer)
If this does not lead to the desired latency, we can think ahead. There are many optimization options such as ONNX.

Cheers :v:

**Reply by frgfm** [reply]
Hello there :wave:

Actually it does, in several situations, since the conda build is way more specific to your system architecture & setup, you can get nice speedups. As a rule of thumb, always favour conda over pip for DL framework installations ;)

**Comment by harindercnvrg** [comment]
@frgfm thanks for sharing these benchmarks. I am having a hard time understanding how exactly you did the installation of doctr using conda because if I do something like ```conda install python-doctr``` the packages installation does not happen.
Can you please provide the steps to reproduce these results you shared using anaconda?

**Reply by felixdittrich92** [reply]
Hi @harindercnvrg 👋 

The benchmark is only about the installed backends via conda vs pip. Doctr has currently no conda build

**Reply by harindercnvrg** [reply]
@felixdittrich92 thanks for the reply. May I ask if you have verified these benchmarks? If so what was the python version you used for anaconda and conda command for installing backend?
